src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -333,6 +341,140 @@ func (r *ClusterReconciler) setLicense(
+		err := r.Get(ctx, key, &node)
+		// if the node is not gone
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf(`unable to fetch node "%s": %w`, pod.Spec.NodeName, err)
+		}
+		// nor has a noexecute taint
+		untainted := true
+		for _, taint := range node.Spec.Taints {
+			if taint.Effect == corev1.TaintEffectNoExecute && taint.Key == corev1.TaintNodeUnreachable {
+				untainted = false
+			}
+		}
+		if untainted {
+			// remove the finalizer and let the pod be restarted
+			if err = r.removePodFinalizer(ctx, pod); err != nil {
+				return fmt.Errorf(`unable to remove finalizer from pod "%s": %w`, pod.Name, err)
+			}
+			continue
+		}
+		// get the node id
+		nodeIDStr, ok := pod.GetAnnotations()[PodAnnotationNodeIDKey]
+		if !ok {
+			return fmt.Errorf("cannot determine node_id for pod %s: %w. not removing finalizer", pod.Name, err)
+		}
+		nodeID, err := strconv.Atoi(nodeIDStr)
+		if err != nil {
+			return fmt.Errorf("node-id annotation is not an integer: %w", err)
+		}
+		// get the broker list
+		redpandaPorts := networking.NewRedpandaPorts(rp)
+		headlessPorts := collectHeadlessPorts(redpandaPorts)
+		clusterPorts := collectClusterPorts(redpandaPorts, rp)
+		headlessSvc := resources.NewHeadlessService(r.Client, rp, r.Scheme, headlessPorts, log)
+		clusterSvc := resources.NewClusterService(r.Client, rp, r.Scheme, clusterPorts, log)
+
+		pki := certmanager.NewPki(r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), clusterSvc.ServiceFQDN(r.clusterDomain), r.Scheme, log)
+
+		adminClient, err := r.AdminAPIClientFactory(ctx, r.Client, rp, headlessSvc.HeadlessServiceFQDN(r.clusterDomain), pki.AdminAPIConfigProvider())
+		if err != nil {
+			return fmt.Errorf("unable to create admin client: %w", err)
+		}
+		brokers, err := adminClient.Brokers(ctx)
+		if err != nil {
+			return fmt.Errorf("unable to fetch brokers: %w", err)
+		}
+		// check if the node in the broker list
+		var broker *admin.Broker
+		for i := range brokers {
+			if brokers[i].NodeID == nodeID {
+				broker = &brokers[i]
+				break
+			}
+		}
+		// if it's not gone
+		if broker != nil {
+			// decommission it
+			if err = adminClient.DecommissionBroker(ctx, nodeID); err != nil {
+				return fmt.Errorf(`unable to decommission node "%d": %w`, nodeID, err)
+			}
+		}
+		//   delete the associated pvc
+		pvc := corev1.PersistentVolumeClaim{}
+		//nolint: gocritic // 248 bytes 6 times is not worth decreasing the readability over
+		for _, v := range pod.Spec.Volumes {
+			if v.PersistentVolumeClaim != nil {
+				key = types.NamespacedName{
+					Name:      v.PersistentVolumeClaim.ClaimName,
+					Namespace: pod.GetNamespace(),
+				}
+			}
+		}
+		err = r.Get(ctx, key, &pvc)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf(`unable to fetch PersistentVolumeClaim "%s/%s": %w`, key.Namespace, key.Name, err)
+			}
+		}
+		if err := r.Delete(ctx, &pvc, &client.DeleteOptions{}); err != nil {
```

Comment:
Should you guard this delete? If `r.Get(ctx, key, &pvc)` will return notFound error in the next reconciliation this call might failed and you will miss the `removePodFinzalizer`